##**Installing PYSPARK**


In [ ]:
pip install pyspark #Installing Pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 281.3 MB 32 kB/s 
     |████████████████████████████████| 199 kB 48.3 MB/s 
  Created wheel for pyspark: filename=pyspark-3.3.0-py2.py3-none-any.whl size=281764026 sha256=0c1f140ea13ab316b92d22546370d859efa703bc0934b6b0eff825bb28547711
  Stored in directory: /root/.cache/pip/wheels/7a/8e/1b/f73a52650d2e5f337708d9f6a1750d451a7349a867f928b885
Successfully built pyspark


In [ ]:
#Initializing PySpark
from pyspark import SparkContext, SparkConf #import sparkcontext and sparkconfig from pyspark
from pyspark.sql import SparkSession #import sparksession from pyspark.sql
from pyspark.sql import SQLContext #import sparkcontext from pyspark.sql

#Spark Config
conf = SparkConf().setAppName("sample_app") #config the app name as variable as conf
sc = SparkContext(conf=conf) # creating the sparkcontext as variable sc
spark = SparkSession.builder.appName('Test').getOrCreate() #creating spark and appname as test to get or create 
sqlContext = SQLContext(sc) # sqlcontext

/usr/local/lib/python3.7/dist-packages/pyspark/sql/context.py:114: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  FutureWarning,


In [ ]:
sc # check the version and app name as sc

<SparkContext master=local[*] appName=sample_app>

##**Read the File**

In [ ]:
#Reading the csv file of the data
olympix= spark.read.format("csv").option("header","true").option("inferSchema","true").option("delimeter",",").load("/content/olympix_data_organized.csv") 
olympix.show() #print output 

+--------------------+---+-------------+----+----------+--------------------+----------+------------+-----------+-----------+
|                name|age|      country|year|Date_Given|              sports|gold_medal|silver_medal|brone_medal|total_medal|
+--------------------+---+-------------+----+----------+--------------------+----------+------------+-----------+-----------+
|      Michael Phelps| 23|United States|2008| 8/24/2008|            Swimming|         8|           0|          0|          8|
|      Michael Phelps| 19|United States|2004| 8/29/2004|            Swimming|         6|           0|          2|          8|
|      Michael Phelps| 27|United States|2012|08-12-2012|            Swimming|         4|           2|          0|          6|
|    Natalie Coughlin| 25|United States|2008| 8/24/2008|            Swimming|         1|           2|          3|          6|
|       Aleksey Nemov| 24|       Russia|2000|10-01-2000|          Gymnastics|         2|           1|          3|     

###**Explanation**

As per the above result, we use spark.read.format and including header and infer schema and delimeter as well.

##**Basic Analysis**

In [ ]:
print(olympix.count()) #counting the total number of rows present in the data set 

8618


###**Explanation**

We can see the above result,we have 8618 rows present in the data set by using count function.

#**Removing the null values**

In [ ]:
# if any row having any Null
# value we are dropping that
# rows
olympix=olympix.dropna(how="all")
olympix.show()

+--------------------+---+-------------+----+----------+--------------------+----------+------------+-----------+-----------+
|                name|age|      country|year|Date_Given|              sports|gold_medal|silver_medal|brone_medal|total_medal|
+--------------------+---+-------------+----+----------+--------------------+----------+------------+-----------+-----------+
|      Michael Phelps| 23|United States|2008| 8/24/2008|            Swimming|         8|           0|          0|          8|
|      Michael Phelps| 19|United States|2004| 8/29/2004|            Swimming|         6|           0|          2|          8|
|      Michael Phelps| 27|United States|2012|08-12-2012|            Swimming|         4|           2|          0|          6|
|    Natalie Coughlin| 25|United States|2008| 8/24/2008|            Swimming|         1|           2|          3|          6|
|       Aleksey Nemov| 24|       Russia|2000|10-01-2000|          Gymnastics|         2|           1|          3|     

###**Explanation**

As per the above output,no null values are present.

##**PySpark code to print the Olympic Sports/games in the dataset.**

In [ ]:
olympix.select("sports").distinct().show() # unique data using distinct function()

+--------------------+
|              sports|
+--------------------+
|          Gymnastics|
|              Tennis|
|              Boxing|
|Short-Track Speed...|
|          Ice Hockey|
|              Rowing|
|                Judo|
|            Softball|
|             Sailing|
|            Swimming|
|       Alpine Skiing|
|          Basketball|
|            Handball|
| Rhythmic Gymnastics|
|            Biathlon|
|           Triathlon|
|           Badminton|
|           Athletics|
|      Figure Skating|
|    Beach Volleyball|
+--------------------+
only showing top 20 rows



###**Explanation**

As per the above output, the list of the sports mentioned in the dataset.

##**PySpark code to the total number of medals in  each Olympic Sport/game**

In [ ]:
res=olympix.groupBy("sports").agg({"total_medal":"sum"}) # grouping the sports nd sum of total medals
res.show() #print output 

+--------------------+----------------+
|              sports|sum(total_medal)|
+--------------------+----------------+
|          Gymnastics|             282|
|              Tennis|              78|
|              Boxing|             188|
|Short-Track Speed...|             138|
|          Ice Hockey|             384|
|              Rowing|             576|
|                Judo|             224|
|            Softball|             134|
|             Sailing|             210|
|            Swimming|             765|
|       Alpine Skiing|              90|
|          Basketball|             287|
|            Handball|             351|
| Rhythmic Gymnastics|              84|
|            Biathlon|             138|
|           Triathlon|              24|
|           Badminton|              96|
|           Athletics|             753|
|      Figure Skating|              54|
|    Beach Volleyball|              48|
+--------------------+----------------+
only showing top 20 rows



In [ ]:
import numpy as np #Importing numpy library
import pandas as pd #Importing pandas library
import matplotlib.pyplot as plt #Importing matplot library
import seaborn as sns # Importing seaborn library

In [ ]:
#converting spark df to pandas df 
pandasDF =res.toPandas() 
print(pandasDF) #print output


                       sports  sum(total_medal)
0                  Gymnastics               282
1                      Tennis                78
2                      Boxing               188
3   Short-Track Speed Skating               138
4                  Ice Hockey               384
5                      Rowing               576
6                        Judo               224
7                    Softball               134
8                     Sailing               210
9                    Swimming               765
10              Alpine Skiing                90
11                 Basketball               287
12                   Handball               351
13        Rhythmic Gymnastics                84
14                   Biathlon               138
15                  Triathlon                24
16                  Badminton                96
17                  Athletics               753
18             Figure Skating                54
19           Beach Volleyball           

##**Sort the result based on the total number of medals.**

In [ ]:
from pyspark.sql.functions import col #importing the col function from pyspark.sql

In [ ]:
olympix.orderBy(col("total_medal").asc(),col("sports").asc()).distinct().show(truncate=False) #sorting the data based on total medals 

+---------------------+---+---------------------+----+----------+-------------------------+----------+------------+-----------+-----------+
|name                 |age|country              |year|Date_Given|sports                   |gold_medal|silver_medal|brone_medal|total_medal|
+---------------------+---+---------------------+----+----------+-------------------------+----------+------------+-----------+-----------+
|Wang Meng            |20 |China                |2006|2/26/2006 |Short-Track Speed Skating|1         |1           |1          |3          |
|Rita König           |23 |Germany              |2000|10-01-2000|Fencing                  |0         |1           |1          |2          |
|Botond Storcz        |25 |Hungary              |2000|10-01-2000|Canoeing                 |2         |0           |0          |2          |
|Ole Einar Bjørndalen |36 |Norway               |2010|2/28/2010 |Biathlon                 |1         |1           |0          |2          |
|Roser Tarragó      

###**Explanation**

As per the above output, the data is sorted according to the total medals won by sports.

##**Total number of medals won by each country in swimming.**

In [ ]:
from pyspark.sql.functions import asc,desc #importing the col function from pyspark.sql

In [ ]:
olympix.filter(olympix.sports=="Swimming").groupBy("country").sum("total_medal").orderBy(desc("sum(total_medal)")).show(10)

+-------------+----------------+
|      country|sum(total_medal)|
+-------------+----------------+
|United States|             267|
|    Australia|             163|
|  Netherlands|              46|
|        Japan|              43|
|       France|              39|
|        China|              35|
|      Germany|              32|
|       Russia|              20|
|        Italy|              16|
|Great Britain|              11|
+-------------+----------------+
only showing top 10 rows



###**Expectation**

As per the above result ,we can see the total number of medals won by each country on swimming 

##**Total number of medals won by each country in Skeleton.**

In [ ]:
olympix.filter(olympix.sports=="Skeleton").groupBy("country").sum("total_medal").orderBy(desc("sum(total_medal)")).show(10) # filtering the skeleton and grouping the country and finding the total medals 

+-------------+----------------+
|      country|sum(total_medal)|
+-------------+----------------+
|       Canada|               4|
|Great Britain|               3|
|  Switzerland|               3|
|United States|               3|
|      Germany|               2|
|       Russia|               1|
|      Austria|               1|
|       Latvia|               1|
+-------------+----------------+



###**Expectation**

As per the above result ,we can see the total number of medals won by each country on skeleton.

##**Number of medals that the US won yearly.**

In [ ]:
olympix.filter(olympix.country=="United States").groupBy("year").agg({"total_medal":"sum"}).show() # filtering the country and grouping the year and finding total medals won

+----+----------------+
|year|sum(total_medal)|
+----+----------------+
|2006|              52|
|2004|             265|
|2012|             254|
|2000|             243|
|2010|              97|
|2008|             317|
|2002|              84|
+----+----------------+



In [ ]:
olympix.filter(olympix.country=="United States").agg({"total_medal":"sum"}).show() #finding the total number of medals won by united states

+----------------+
|sum(total_medal)|
+----------------+
|            1312|
+----------------+



###**Explanation**

As per the above result, we can see how many medals won by united states on participated year.

The total number of medals won by united states is **1312**.

##**Total number of medals won by each country.**

In [ ]:
res1 = olympix.groupBy("country").agg({"total_medal":"sum"}) # grouping the country and sum of total medals 
res1.show() #print output 

+--------------+----------------+
|       country|sum(total_medal)|
+--------------+----------------+
|        Russia|             768|
|      Paraguay|              17|
|Chinese Taipei|              20|
|        Sweden|             181|
|       Eritrea|               1|
|     Singapore|               7|
|      Malaysia|               3|
|        Turkey|              28|
|       Germany|             629|
|   Afghanistan|               2|
|         Sudan|               1|
|        France|             318|
|        Greece|              59|
|     Sri Lanka|               1|
|       Algeria|               8|
|          Togo|               1|
|      Slovakia|              35|
|     Argentina|             141|
|       Belgium|              18|
| Great Britain|             322|
+--------------+----------------+
only showing top 20 rows



##**Oldest athlete in the olympics**

In [ ]:
olympix.select("name","age").sort(desc("age")).show(5) #finding the oldest age of athletes

+--------------------+---+
|                name|age|
+--------------------+---+
|          Ian Millar| 61|
|           Mark Todd| 56|
|            Mac Cone| 55|
|        Nick Skelton| 54|
|Lesley Thompson-W...| 52|
+--------------------+---+
only showing top 5 rows



###**Explanaton**

As per the above result the oldest athlete is **IAN MILLAR** and his age is **61**.

##**Which country was he/she from?**

In [ ]:
olympix.select("name","age","country").sort(desc("age")).show(5) #finding the country of oldest athletes

+--------------------+---+-------------+
|                name|age|      country|
+--------------------+---+-------------+
|          Ian Millar| 61|       Canada|
|           Mark Todd| 56|  New Zealand|
|            Mac Cone| 55|       Canada|
|        Nick Skelton| 54|Great Britain|
|Lesley Thompson-W...| 52|       Canada|
+--------------------+---+-------------+
only showing top 5 rows



###**Explanation**

As per the above result,the oldest athlete is IAN MILLER and he belongs to **CANADA**